api/odl.py

```diff
@@ -786,6 +765,82 @@ class ODLImporter(OPDSImporter):
     # about the license.
     LICENSE_INFO_DOCUMENT_MEDIA_TYPE = 'application/vnd.odl.info+json'
 
+    @classmethod
+    def parse_license(
+            cls,
+            identifier: str,
+            total_checkouts: Optional[int],
+            concurrent_checkouts: Optional[int],
+            expires: Optional[datetime.datetime],
+            checkout_link: Optional[str],
+            odl_status_link: Optional[str],
+            do_get: Callable
+    ) -> Optional[LicenseData]:
+        remaining_checkouts = None
+        available_concurrent_checkouts = None
+
+        while True:
+            if total_checkouts is not None:
+                total_checkouts = int(total_checkouts)
+
+                if total_checkouts <= 0:
+                    logging.info(
+                        f"License # {identifier} expired since "
+                        f"the total number of checkouts is {total_checkouts}"
+                    )
+                    break
+
+            if expires:
+                if not isinstance(expires, datetime.datetime):
+                    expires = dateutil.parser.parse(expires)
+
+                expires = util.datetime_helpers.to_utc(expires)
+                now = util.datetime_helpers.utc_now()
+
+                if expires <= now:
+                    logging.info(
+                        f"License # {identifier} expired at {expires} (now is {now})"
+                    )
+                    break
+
+            if odl_status_link:
+                status_code, _, response = do_get(
+                    odl_status_link, headers={}
+                )
+
+                if status_code < 400:
+                    status = json.loads(response)
+                    checkouts = status.get("checkouts", {})
+                    remaining_checkouts = checkouts.get("left")
+                    available_concurrent_checkouts = checkouts.get("available")
+
+            if remaining_checkouts is None:
```

Comment:
I think that if `left` property is missing but `total_checkouts` is there it means that nobody used it so far.
But on the other hand it might mean that it actually expired and lack of `left` means that no checkouts left.

Reply:
If `left` is missing, it is an error according to the [ODL spec](https://drafts.opds.io/odl-1.0#41-syntax). Have we seen that case?

Reply:
@tdilauro, I even found [the case](https://license.feedbooks.net/copy/status/?uuid=1d9d8404-4e54-44fb-8494-cbef07c66b3d) in the DPLA Exchange Feed when `left` is missing.

Reply:
I put this question to De Marque. Hopefully will have a response soon.

Reply:
De Marque indicated that you should not expect to see `left` when there is an unlimited number of checkouts (which is not consistent with the spec). In that case, `total_checkouts` should be missing also, as it is in [the example](https://license.feedbooks.net/copy/status/?uuid=1d9d8404-4e54-44fb-8494-cbef07c66b3d) you gave above.

Have you seen any cases where `left` is missing, but `total_checkouts` is present?

Reply:
@tdilauro, yes, there are cases like that in the DPLA Exchange feed. I added a new cell to [the Jupyter notebook](https://colab.research.google.com/drive/1mLSUuNj0MFkzeIayDp2YPhd0Jq0OK-BO?usp=sharing) where all the licenses like that are shown.